[Rachel Chua] ip

RachelChua opened the pull request:
# Duke 
> "For every minute spent organising, an hour is earned".-[Benjamin Franklin](https://medium.com/@steveagyeibeyondlifestyle/for-every-minute-spent-in-organizing-an-hour-is-earned-969e3ca0627c)

Duke helps organise your life. It's 
- fun
- easy to use
- text-based

All you need to do is 
1. install it 
2. add your tasks
3. let it help you organise your life :smiley:

And it is **Free**!

Some of the features are:
- [x] Add tasks: todos, deadline, events
- [x] Find tasks
- [x] Delete tasks
- [x] Show list of tasks
- [x] Mark task as done

Here is a peak of the main method 
```
 public static void main(String[] args) {
        Application.launch(Duke.class, args);
    }
```

Commit:
Add Gradle support

RachelChua committed:
Level 1

RachelChua committed:
Level 2

RachelChua committed:
Level 3

RachelChua committed:
Level 4

RachelChua committed:
Level 6

RachelChua committed:
Automated text UI testing and level 5

RachelChua committed:
Automated text UI testing and level 5

RachelChua committed:
test

RachelChua committed:
test1

RachelChua committed:
test branching

RachelChua committed:
Store deadline dates as a java.time.LocalDate

Duke stores deadline dates as a java.time.LocalDate instead of a String.

Accept dates in a format such as yyyy-mm-dd format (e.g., 2019-10-15) and print in a different format such as MMM dd yyyy e.g., (Oct 15 2019).

RachelChua committed:
Save tasks in hard disk automatically when adding task.

This is such that the data will remember the task that is previously
added to the list

RachelChua committed:
Merge branch 'branch-Level-8'

RachelChua committed:
Test merging

RachelChua committed:
Merge branch 'branch-Level-7'

RachelChua committed:
Add Ui class to deal with interaction with users

RachelChua committed:
Add Parser, Storage and TaskList class for OOP.

TaskList deals with operation to perform on tasks in the list
Parser deals with making sense of the user command
Storage deals with loading tasks from file

RachelChua committed:
Update mark and unmark operation to update file saved in hard disk

When mark and unmark functions are called, the file in hard disk
will also be updated accordingly such that the next time we load the
tasks from the file, the mark or unmark symbol will be correct

RachelChua committed:
Update classes to make use of the Ui class

RachelChua committed:
Update Duke class to make use of classes for better OOP.

RachelChua committed:
Update Duke class to make use of OOP. Make minor changes to
Events class, TaskList class.

RachelChua committed:
Use variables to reduce code repetition, add modifyFileHelper needed in
mark and unmark function

RachelChua committed:
Add all classes into duke package

RachelChua committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

RachelChua committed:
Update such that when exit program, it generates list and writes as new
file instead of writing into the file everytime for add delete mark or
unmark operation

RachelChua committed:
Add Command class for more OOP

RachelChua committed:
Add JUnit test to test behaviour of toString method in Todo class and
Event class

RachelChua committed:
Create a JAR file for easy distribution

RachelChua committed:
Test commit

RachelChua committed:
Add JavaDocs to some classes.

RachelChua committed:
Add find command to list out task with the keyword after find

RachelChua committed:
Follow the coding standard

RachelChua committed:
Test commit

RachelChua committed:
Test commit

RachelChua committed:
Merge branch 'branch-Level-9'

RachelChua committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Deadline.java
#	src/main/java/duke/DeadlineCommand.java
#	src/main/java/duke/Event.java
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Task.java
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/ToDo.java

RachelChua committed:
Try to fix merge conflict

RachelChua committed:
Fix merge conflict

RussellDash332 commented on DATA/duke.txt:
```diff
@@ -0,0 +1,12 @@
+[T][] 1
+[T][X] 1
+[T][] 3
+[T][] 1
+[T][] 2
+[T][] book 
+[T][] book read
+[D][] study book (by: Aug 20 2002)
+[T][] 100
+[D][X] test (by: Feb 22 2022)
+[E][] outing (at: 2pm)
+[T][] 1
```
Perhaps including this in `.gitignore` would be a great idea since it's a data text and it's by right a private file.

RussellDash332 commented on src/main/java/duke/DeadlineCommand.java:
```diff
@@ -0,0 +1,28 @@
+
+    @Override
+    void execute(String fullCommand, ArrayList<Task> listOfTasks, Ui ui, Storage storage) throws IOException, DukeDeadlineEmptyException {
+        if(fullCommand.length() == 8){
```
Do you think there should have been whitespaces to separate the `if`, the condition, and the curly brace?

RussellDash332 commented on src/main/java/duke/DeadlineCommand.java:
```diff
@@ -0,0 +1,28 @@
+        TaskList taskList = new TaskList(listOfTasks);
+        try {
+            LocalDate date = LocalDate.parse(subString);
+            Task t = new Deadline(subS,false, date);
```
Small issue but there seems to be a missing space here (and line 24 as well) 😄

RussellDash332 commented on src/main/java/duke/DeleteCommand.java:
```diff
@@ -0,0 +1,12 @@
+
+public class DeleteCommand extends Command {
+    @Override
+     void execute(String taskName, ArrayList<Task> listOfTasks, Ui ui, Storage storage) throws IOException {
```
Should this have one less space?

RussellDash332 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,78 @@
+     * @return Command depending on the substrings of different commands.
+     * @throws DukeUnknownWordException
+     */
+    Command parse() throws DukeUnknownWordException {
+        if (isSubstringForDeleteCommand()){
+            return new DeleteCommand();
+        } else if (isSubstringForMarkCommand()){
+            return new MarkCommand();
+        } else if (isSubstringForUnmarkCommand()){
+            return new UnmarkCommand();
+        } else if (isSubstringForDeadlineCommand()){
+            return new DeadlineCommand();
+        } else if (isSubstringForEventCommand()) {
+            return new EventCommand();
+        } else if (isSubstringForTodoCommand()){
+            return new TodoCommand();
+        } else if (isSubstringForShowListCommand()){
+            return new ShowListCommand();
+        } else if (isSubstringForFindCommand()) {
+            return new FindCommand();
+            //} else if (isSubStringForExitCommand()){
+            //  return new ExitCommand();
+        } else {
+            throw new DukeUnknownWordException();
+        }
+    }
```
I like how you decided to create a separate command class for each command! This makes the abstraction clearer and easier to maintain as new commands are added in the future 🥳

RussellDash332 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,78 @@
+
+    boolean isSubStringForExitCommand() {
+        return this.stringToParse.equals("bye");
+    }
+    boolean isSubstringForShowListCommand() {
```
I would suggest to separate these methods by a blank line 👍🏻

RussellDash332 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,107 @@
+ * Gets tasks from file, creates a new file if file is absent, and upload tasks in file.
+ */
+public class Storage {
+    private String pathname;
```
Do you think we can use `pathName` instead?

RussellDash332 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,71 @@
+package duke;
+import java.util.ArrayList;
+
+public class Ui {
```
I also like the abstraction of the UI methods here, well done ⭐

RussellDash332 reviewed (commented):
LGTM! The few nits to deal with in my opinion are just the capitalization and spacing issues, this can be resolved by doing `A-CheckStyle` as I haven't seen any tags on this at the moment of writing.

One thing in particular that I'm delighted about is the noticeable abstraction here. You seem to put different commands in separate classes, as well as different types of exceptions as well. I can see there's a lot of effort put in here, good job! 😃

RachelChua committed:
Implement GUI to Duke

RachelChua committed:
Add greet Ui to work with GUI and add padding for text in GUI

RachelChua committed:
Add exit method to end the program when user inputs bye

alvintfl commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,25 @@
+import java.util.ArrayList;
+
+/**
+ * Reperesents the commands inputted by users.
```
Perhaps you made a typo and it should be Represents instead?

alvintfl commented on src/main/java/duke/FindCommand.java:
```diff
@@ -0,0 +1,20 @@
+               // System.out.println(count + "." + listOfTasks.get(i).toString());
+                 s = s + "\n" + count + "." + listOfTasks.get(i).toString();
+            } else {
+            }
```
Perhaps there is no need for the else brace if it does nothing?

alvintfl commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,107 @@
+                } else {
+                    int openBracketIndex = string.indexOf("(");
+                    int closeBracketIndex = string.indexOf(")");
+                    String nameOfMarkedTaskForDeadlineOrEventTask = string.substring(7, openBracketIndex - 1);
```
Maybe ```nameOfMarkedTaskForDeadlineOrEventTask``` can be renamed to something shorter like ```markedDeadlineOrEventName``` so that it is easier to read?

alvintfl commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,78 @@
+     * @return Command depending on the substrings of different commands.
+     * @throws DukeUnknownWordException
+     */
+    Command parse() throws DukeUnknownWordException {
+        if (isSubstringForDeleteCommand()){
+            return new DeleteCommand();
+        } else if (isSubstringForMarkCommand()){
+            return new MarkCommand();
+        } else if (isSubstringForUnmarkCommand()){
+            return new UnmarkCommand();
+        } else if (isSubstringForDeadlineCommand()){
+            return new DeadlineCommand();
+        } else if (isSubstringForEventCommand()) {
+            return new EventCommand();
+        } else if (isSubstringForTodoCommand()){
+            return new TodoCommand();
+        } else if (isSubstringForShowListCommand()){
+            return new ShowListCommand();
+        } else if (isSubstringForFindCommand()) {
+            return new FindCommand();
+       // } else if (isSubStringForExitCommand()){
+         //   return new ExitCommand();
+        } else {
+            throw new DukeUnknownWordException();
+        }
+    }
```
I like how you put the checking of a command in other methods to make ```parse``` easier to read 👍

alvintfl commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,107 @@
+                    String atOrByInfo = string.substring(openBracketIndex + 5, closeBracketIndex);
+
+                    if (typeOfTask.equals("D")) {
+                        if (gotMarkSymbol) {
+                            listStored.add(new Deadline(nameOfMarkedTaskForDeadlineOrEventTask,
+                                    true, atOrByInfo));
+                        } else {
+                            listStored.add(new Deadline(nameOfUnmarkedTaskForDeadlineOrEventTask,
+                                    false, atOrByInfo));
+                        }
+                    } else if (typeOfTask.equals("E")) {
+                        if (gotMarkSymbol) {
+                            listStored.add(new Event(nameOfMarkedTaskForDeadlineOrEventTask,
+                                    true, atOrByInfo));
+                        } else {
+                            listStored.add(new Event(nameOfUnmarkedTaskForDeadlineOrEventTask,
+                                    false, atOrByInfo));
+                        }
```
Maybe we can reduce the nesting needed if we pass ```gotMarkSymbol``` as an argument?

alvintfl reviewed (commented):
Overall, the quality of the code is good! Just felt some parts could be simplified further. 
LGTM. 💯

RachelChua committed:
Remove some files of GUI

RachelChua committed:
Improve code quality

RachelChua committed:
Use assertions

Assertions are used to document important assumptions that should hold
in the code.

RachelChua committed:
Merge pull request #1 from RachelChua/branch-A-CodeQuality

Improve code quality

RachelChua committed:
Merge branch 'master' of https://github.com/RachelChua/ip

RachelChua committed:
Merge branch 'master' into branch-A-Assertions

RachelChua committed:
Merge pull request #2 from RachelChua/branch-A-Assertions

Use assertions

RachelChua committed:
Merge branch 'master' of https://github.com/RachelChua/ip

RachelChua committed:
Deal with duplicate items

Duplicate items are previously unable to be detected. They can be added
into list again, which is not ideal.

New feature is added such that task with duplicate names
will not be added since it is already in the list.

RachelChua committed:
Change code to improve code quality

RachelChua committed:
Add user guide and UI.png

RachelChua committed:
Update README.md
